xmtp_api_d14n/src/compat/d14n.rs

```diff
@@ -41,80 +48,320 @@ where
+                                created_ns: unsigned_originator_envelope.originator_ns as u64,
+                                group_id: request.group_id.clone(),
+                                data: v1_group_message.data,
+                                sender_hmac: v1_group_message.sender_hmac,
+                            })),
+                        }));
+                    }
+                }
+
+                Some(Err(ApiError::<E>::ProtoError(ProtoError::NotFound(
+                    "Invalid Payload Type".into(),
+                ))))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Ok(QueryGroupMessagesResponse {
+            messages,
+            paging_info: None,
+        })
     }
     async fn query_welcome_messages(
         &self,
         request: QueryWelcomeMessagesRequest,
     ) -> Result<QueryWelcomeMessagesResponse, Self::Error> {
-        todo!()
+        let query = EnvelopesQuery {
+            topics: vec![build_welcome_message_topic(
+                request.installation_key.as_slice(),
+            )],
+            originator_node_ids: Vec::new(), // todo: set later
+            last_seen: None,                 // todo: set later
+        };
+
+        let response_envelopes = QueryEnvelopes::builder()
+            .envelopes(query)
+            .limit(
+                request
+                    .paging_info
+                    .map_or(DEFAULT_PAGINATION_LIMIT, |paging| paging.limit),
+            )
+            .build()
+            .unwrap()
+            .query(&self.message_client)
+            .await?;
+
+        let messages = response_envelopes
+            .envelopes
+            .into_iter()
+            .filter_map(|envelope| {
+                let unsigned_originator_envelope =
+                    extract_unsigned_originator_envelope(&envelope).ok()?;
+                let client_envelope = extract_client_envelope(&envelope).ok()?;
+                let payload = client_envelope.payload?;
+
+                if let Payload::WelcomeMessage(welcome_message) = payload {
+                    if let Some(welcome_message_input::Version::V1(v1_welcome_message)) =
+                        welcome_message.version
+                    {
+                        return Some(Ok(WelcomeMessage {
+                            version: Some(welcome_message::Version::V1(welcome_message::V1 {
+                                id: unsigned_originator_envelope.originator_sequence_id,
+                                created_ns: unsigned_originator_envelope.originator_ns as u64,
+                                installation_key: request.installation_key.clone(),
+                                data: v1_welcome_message.data,
+                                hpke_public_key: v1_welcome_message.hpke_public_key,
+                            })),
+                        }));
+                    }
+                }
+
+                Some(Err(ApiError::<E>::ProtoError(ProtoError::NotFound(
```

Comment:
Do we want to halt on a single bad payload, or should we ignore it?

Reply:
Good catch, makes sense, will fix it.